File: leetcode/py/p50-pow-x-n.py
# LeetCode 50. Pow(x, n)


def exp_by_squaring_positive(x: float, n: int):
    """Exponent by squaring algorithm, with check for -x already done."""
    if n == 0:
        return 1.0
    if n % 2 == 0:
        return exp_by_squaring_positive(x * x, n // 2)
    return x * exp_by_squaring_positive(x * x, (n - 1) // 2)


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1.0
        if x == 0:
            return 0.0
        if n < 0:
            x = 1 / x
            n = -n
        return exp_by_squaring_positive(x, n)


test_cases = (
    (2.0, 10, 1024.0),
    (2.1, 3, 9.261),
    (2.0, -2, 0.25),
    (0.0, -2, 0.0),
    (0.0, 2, 0.0),
    (0.0, 0, 1.0),
)

solution = Solution()
for x, n, exp in test_cases:
    act = solution.myPow(x, n)
    if exp != act:
        print(f"fail, x: {x}, n: {n}, exp: {exp}, act: {act}")

File: leetcode/cpp/p50-pow-x-n.cpp
// LeetCode p50. Pow(x, n).

#include <iostream>

// Exponent by squaring algorithm, with check for -x already done.
double exp_by_squaring_positive(double x, int64_t n)
{
    if (n == 0) {
        return 1.0;
    }
    if (n % 2 == 0) {
        return exp_by_squaring_positive(x * x, n / 2);
    }
    return x * exp_by_squaring_positive(x * x, (n - 1) / 2);
}

class Solution {
public:
    double myPow(double x, int n)
    {
        if (n == 0) {
            return 1.0;
        }
        if (x == 0.0) {
            return 0.0;
        }
        if (n < 0) {
            int64_t n64 = n;
            return exp_by_squaring_positive(1 / x, -n64);
        }
        return exp_by_squaring_positive(x, n);
    }
};

int main()
{
    struct TestCase {
        double x;
        int n;
        double exp;
    };

    const TestCase test_cases[] = {
        { 2.0, 10, 1024.0 },
        { 2.1, 3, 9.261 },
        { 2.0, -2, 0.25 },
        { 0.0, 0, 1.0 },
        { 2.0, 0, 1.0 },
        { -2.0, 0, 1.0 },
        { -2.0, 1, -2.0 },
    };

    Solution solution;
    for (const auto& tc : test_cases) {
        const auto act = solution.myPow(tc.x, tc.n);
        if (tc.exp != act) {
            std::cerr << "Fail, x: " << tc.x
                      << ", n: " << tc.n
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
}